### PR TITLE
[Appkit] Bindings - part 2

### PR DESCRIPTION
**Pull request by prmallic**
Rebased on PR#7187.

### CODE REVIEW

File: src/appkit.cs [hunk 1]
```diff
@@ -3062,6 +3065,7 @@ partial interface NSCollectionViewDelegate {
 		bool CanDragItems (NSCollectionView collectionView, NSIndexSet indexes, NSEvent evt);
 
 		[Export ("collectionView:writeItemsAtIndexes:toPasteboard:")]
+		[Deprecated (PlatformName.MacOSX, 10, 15, message: "Use the 'GetPasteboardWriter' method instead")]
```
Comment (mandel-macaque) [inline src/appkit.cs]: Missin '.' at the end of the sentence.

File: src/appkit.cs [hunk 2]
```diff
@@ -3087,6 +3091,7 @@ partial interface NSCollectionViewDelegate {
 
 		[Mac (10,11)]
 		[Export ("collectionView:writeItemsAtIndexPaths:toPasteboard:")]
+		[Deprecated (PlatformName.MacOSX, 10, 15, message: "Use the 'GetPasteboardWriter' method instead")]
```
Comment (mandel-macaque) [inline src/appkit.cs]: same

File: src/appkit.cs [hunk 3]
```diff
@@ -17182,6 +17198,7 @@ interface NSTableViewDataSource {
 		void SortDescriptorsChanged (NSTableView tableView, NSSortDescriptor [] oldDescriptors);
 	
 		[Export ("tableView:writeRowsWithIndexes:toPasteboard:")]
+		[Deprecated (PlatformName.MacOSX, 10, 15, message: "Use the 'GetPasteboardWriterForRow' method instead")]
```
Comment (mandel-macaque) [inline src/appkit.cs]: same, .

File: src/appkit.cs [hunk 4]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+	}
+
+	[Mac (10,15)]
+	[BaseType (typeof(NSObject))]
+	[DisableDefaultCtor]
+	interface NSCollectionLayoutAnchor : NSCopying, INSCopying
+	{
+		[Static]
+		[Export ("layoutAnchorWithEdges:")]
+		NSCollectionLayoutAnchor LayoutAnchorWithEdges (NSDirectionalRectEdge edges);
```
Comment (mandel-macaque) [inline src/appkit.cs]: Not needed 'WithEdeges' it can be inferred from the parameter.

File: src/appkit.cs [hunk 5]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+
+		[Static]
+		[Export ("layoutAnchorWithEdges:absoluteOffset:")]
+		NSCollectionLayoutAnchor LayoutAnchorWithEdgesAbsoluteOffset (NSDirectionalRectEdge edges, CGPoint absoluteOffset);
```
Comment (mandel-macaque) [inline src/appkit.cs]: Same, remove WithEdgesAbsoluteOffset.

File: src/appkit.cs [hunk 6]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+
+		[Static]
+		[Export ("layoutAnchorWithEdges:fractionalOffset:")]
+		NSCollectionLayoutAnchor LayoutAnchorWithEdgesFractionalOffset (NSDirectionalRectEdge edges, CGPoint fractionalOffset);
```
Comment (mandel-macaque) [inline src/appkit.cs]: So, following my above comment, we have the factional and the abolsute offset. I would add that as an enum, remove the WithEdges and do a single constructor that takes the enum and does the correct call.

File: src/appkit.cs [hunk 7]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+	}
+
+	[Mac (10,15)]
+	[BaseType (typeof(NSObject))]
+	[DisableDefaultCtor]
+	interface NSCollectionLayoutDimension : NSCopying, INSCopying
+	{
+		[Static]
+		[Export ("fractionalWidthDimension:")]
+		NSCollectionLayoutDimension CreateFractionalWidthDimension (nfloat fractionalWidth);
```
Comment (mandel-macaque) [inline src/appkit.cs]: Remove WithDimension

File: src/appkit.cs [hunk 8]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+
+		[Static]
+		[Export ("fractionalHeightDimension:")]
+		NSCollectionLayoutDimension CreateFractionalHeightDimension (nfloat fractionalHeight);
```
Comment (mandel-macaque) [inline src/appkit.cs]: Remove with dimension. Following the comment above, enum usage and a single method?

File: src/appkit.cs [hunk 9]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+
+		[Static]
+		[Export ("absoluteDimension:")]
+		NSCollectionLayoutDimension CreateAbsoluteDimension (nfloat absoluteDimension);
+
+		[Static]
+		[Export ("estimatedDimension:")]
+		NSCollectionLayoutDimension CreateEstimatedDimension (nfloat estimatedDimension);
```
Comment (mandel-macaque) [inline src/appkit.cs]: Ditto.

File: src/appkit.cs [hunk 10]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+
+		[Export ("numberOfOptions")]
+		nint NumberOfOptions { get; set; }
+
+		[Export ("setImage:atIndex:")]
+		void SetImage ([NullAllowed] NSImage image, nint index);
+
+		[Export ("imageAtIndex:")]
+		[return: NullAllowed]
+		NSImage ImageAtIndex (nint index);
```
Comment (rolfbjarne) [inline src/appkit.cs]: This has a verb, sounds more like C#, and is also symmetrical to `SetImage`:

```suggestion
		NSImage GetImage (nint index);
```

File: src/appkit.cs [hunk 11]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+
+		[Export ("setLabel:atIndex:")]
+		void SetLabel (string label, nint index);
+
+		[Export ("labelAtIndex:")]
+		[return: NullAllowed]
+		string LabelAtIndex (nint index);
```
Comment (rolfbjarne) [inline src/appkit.cs]: Same:

```suggestion
		string GetLabel (nint index);
```

File: src/appkit.cs [hunk 12]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+		Selector Action { get; set; }
+
+		[Export ("enabled")]
+		bool Enabled { [Bind ("isEnabled")] get; set; }
+
+		[Export ("setEnabled:atIndex:")]
+		void SetEnabled (bool enabled, nint index);
+
+		[Export ("isEnabledAtIndex:")]
+		bool IsEnabledAtIndex (nint index);
```
Comment (rolfbjarne) [inline src/appkit.cs]: Same:

```suggestion
		bool GetEnabled (nint index);
```

File: src/appkit.cs [hunk 13]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+
+		// @property (readwrite, copy) NSString * _Null_unspecified customizationLabel;
+		[Export ("customizationLabel")]
```
Comment (rolfbjarne) [inline src/appkit.cs]: ArgumentSemantic.Copy is important, and shouldn't be skipped (the `copy` part from the Objective-C property declaration above)

```suggestion
		[Export ("customizationLabel", ArgumentSemantic.Copy)]
```

File: src/appkit.cs [hunk 14]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+	[BaseType (typeof(NSObject))]
+	[DisableDefaultCtor]
+	interface NSCollectionLayoutSize : NSCopying, INSCopying
+	{
+		[Static]
+		[Export ("sizeWithWidthDimension:heightDimension:")]
+		NSCollectionLayoutSize Create (NSCollectionLayoutDimension width, NSCollectionLayoutDimension height);
+
+		[Export ("widthDimension")]
+		NSCollectionLayoutDimension Width { get; }
```
Comment (rolfbjarne) [inline src/appkit.cs]: I think `WidthDimension` sounds just as good as `Width`, and it has the advantage of being identical to Swift/Objective-C, so I would go with `WidthDimension` here.

```suggestion
		NSCollectionLayoutDimension WidthDimension { get; }
```

File: src/appkit.cs [hunk 15]
```diff
@@ -26692,4 +26729,338 @@ interface NSWindowTabGroup
...
+
+		[Export ("heightDimension")]
+		NSCollectionLayoutDimension Height { get; }
```
Comment (rolfbjarne) [inline src/appkit.cs]: Same:

```suggestion
		NSCollectionLayoutDimension HeightDimension { get; }
```

File: src/AppKit/NSCollectionLayoutAnchor.cs [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+namespace AppKit {
+
+	[Mac (10, 15)]
+	public enum NSCollectionLayoutAnchorOffsetType {
+		Absolute,
+		Fractional,
+	}
+
+	public partial class NSCollectionLayoutAnchor {
+		public static NSCollectionLayoutAnchor GetLayoutAnchor (NSDirectionalRectEdge edges, NSCollectionLayoutAnchorOffsetType offsetType, CGPoint offset) =>
```
Comment (spouliot) [inline src/AppKit/NSCollectionLayoutAnchor.cs]: In general (I did not read the associated doc) when type 'X' has static members that returns 'X' then it's more a `Create` than a `Get`, e.g.

```csharp
public class Url {
   // good
   static Url Create (string url) { ... }
   // bad
   static Url Get (string url) { ... }

   // good
   static string GetDomain (Url url) { ... }
   // bad
   static string CreateDomain (Url url) { ... }
}
```

File: src/appkit.cs [hunk 17]
```diff
@@ -26850,12 +26856,211 @@ public enum NSToolbarItemGroupControlRepresentation : long
...
+	}
+
+	[Mac (10,15)]
+	[BaseType (typeof(NSTouchBarItem))]
+	[DisableDefaultCtor]
+	interface NSPickerTouchBarItem
+	{
+		[Export ("initWithIdentifier:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (string identifier);
```
Comment (spouliot) [inline src/appkit.cs]: should not this be `NSTouchBarItemIdentifier` (not `string`) ?

File: src/frameworks.sources [hunk 18]
```diff
@@ -1512,6 +1514,7 @@ UIKIT_CORE_SOURCES = \
     UIKit/UIPrintInteractionController.cs \
     UIKit/UIStringAttributes.cs \
     UIKit/UITypes.cs \
+	XKit/Types.cs \
```
Comment (spouliot) [inline src/frameworks.sources]: misaligned - tabs vs spaces?

**Comment by prmallic** [comment]
So removing the above .ctor (`[DesignatedInitializer]`) from `NSPickerTouchBarItem` and adding the skip case to the test. Thanks @dalexsoto!

File: tests/introspection/ApiCtorInitTest.cs [hunk 19]
```diff
@@ -477,6 +477,11 @@ protected virtual bool Match (ConstructorInfo ctor, Type type)
 			case "UICollectionViewCompositionalLayout":
 				// Explicitly disabled ctors - (instancetype)init NS_UNAVAILABLE;
 				return true;
+			case "NSPickerTouchBarItem":
+				if (ctor.ToString () == $"Void .ctor(System.String)")
```
Comment (spouliot) [inline tests/introspection/ApiCtorInitTest.cs]: please add a comment why this ctor is being disabled (like the above cases already have)

there's nothing in `/Applications/Xcode112-beta2.app//Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/AppKit.framework/Versions/C/Headers/NSPickerTouchBarItem.h` to suggest it should not be available.

**Comment by prmallic** [comment]
Also, the Xtro here generates double names

**Pull request merged**